pkg/cli/gettoken/gettoken.go

```diff
@@ -0,0 +1,269 @@
+	}
+
+	return nil
+}
+
+func (o *GetTokenOptions) Validate() error {
+	if o.IssuerURL == "" {
+		return fmt.Errorf("--issuer-url is required")
+	}
+	if o.ClientID == "" {
+		return fmt.Errorf("--client-id is required")
+	}
+
+	return nil
+}
+
+// Run starts the authentication flow with a caching and refreshing capability.
+// If refresh token is found, it tries to use it to get a valid id token from
+// external OIDC issuer. If not, it forces user to log in.
+func (o *GetTokenOptions) Run() error {
+	tokenCacheKey := tokencache.Key{
+		IssuerURL:      o.IssuerURL,
+		ClientID:       o.ClientID,
+		ExtraScopes:    o.ExtraScopes,
+		CACertFilename: o.CACertFilename,
+		SkipTLSVerify:  o.InsecureTLS,
+	}
+
+	tokenSet, err := o.tokenCache.FindByKey(o.tokenCacheDir, tokenCacheKey)
+	if err != nil {
+		return err
+	}
+	alreadyValid, idToken, refreshToken, err := o.getToken(context.TODO(), tokenSet)
+	if err != nil {
+		return err
+	}
+
+	idTokenClaims, err := jwt.DecodeWithoutVerify(idToken)
+	if err != nil {
+		return fmt.Errorf("you got an invalid token: %w", err)
+	}
+
+	if !alreadyValid {
+		err = o.tokenCache.Save(o.tokenCacheDir, tokenCacheKey, tokencache.Set{
+			IDToken:      idToken,
+			RefreshToken: refreshToken,
+		})
+		if err != nil {
+			return fmt.Errorf("could not write into the token cache")
+		}
+	}
+
+	if err := o.credWriter.Write(idToken, idTokenClaims.Expiry); err != nil {
+		return fmt.Errorf("could not write the token to client-go: %w", err)
+	}
+	return nil
+}
+
+func (o *GetTokenOptions) getToken(ctx context.Context, cache *tokencache.Set) (bool, string, string, error) {
+	if cache != nil {
+		claims, err := jwt.DecodeWithoutVerify(cache.IDToken)
+		if err != nil {
+			return false, "", "", fmt.Errorf("invalid token cache (you may need to remove): %w", err)
+		}
+		if !claims.IsExpired(time.Now()) {
+			return true, cache.IDToken, cache.RefreshToken, nil
+		}
+	}
+
+	if cache != nil && cache.RefreshToken != "" {
+		idToken, refreshToken, err := o.authenticator.Refresh(ctx, cache.RefreshToken)
+		if err == nil {
+			return false, idToken, refreshToken, nil
+		}
+	}
+
+	idToken, refreshToken, err := o.doAuthCode(ctx)
+	if err != nil {
+		return false, "", "", err
+	}
+
+	return false, idToken, refreshToken, err
+}
+
+// doAuthCode does the auth code flow with PKCE(if the issuer supports it).
+func (o *GetTokenOptions) doAuthCode(ctx context.Context) (string, string, error) {
+	state, err := oidc.NewState()
+	if err != nil {
+		return "", "", fmt.Errorf("could not generate a state: %w", err)
+	}
+	nonce, err := oidc.NewNonce()
+	if err != nil {
+		return "", "", fmt.Errorf("could not generate a nonce: %w", err)
+	}
+	p, err := pkce.New(o.authenticator.SupportedPKCEMethods())
+	if err != nil {
+		return "", "", fmt.Errorf("could not generate PKCE parameters: %w", err)
+	}
+	input := oidc.GetTokenByAuthCodeInput{
+		BindAddress: o.CallbackAdresses,
+		State:       state,
+		Nonce:       nonce,
+		PKCEParams:  p,
+	}
```

Comment:
Yes, you are right, they are not used directly in here. But just to have a similarity with the function that is copied from https://github.com/int128/kubelogin/blob/79a6f8596774d9d1fff2319ddc84a3a87f80fa48/pkg/usecases/authentication/authcode/browser.go#L34-L96 to track the changes, at least for a while. Otherwise, if there is a hidden bug, we wouldn't compare the code with upstream plugin and ensure whether bug is fixed in there.

Reply:
hide it inside of the OAuth2 request implementation

Reply:
Moved inside OAuth2 request implementation.